### PR TITLE
add whole-region mmap area for vfu_setup_region()

### PR DESCRIPTION
**Pull request by jlevon**
If an fd is provided, automatically add a mmap area covering the entire region
unless an mmap_areas array is provided.

Signed-off-by: John Levon <john.levon@nutanix.com>

### CODE REVIEW

File: lib/libvfio-user.c [hunk 1]
```diff
@@ -151,14 +150,15 @@ dev_get_caps(vfu_ctx_t *vfu_ctx, vfu_reg_info_t *vfu_reg, bool is_migr_reg,
...
         for (i = 0; i < nr_mmap_areas; i++) {
-            (*fds)[i] = vfu_reg->fd;
-            sparse->areas[i].offset = (__u64)mmap_areas->areas[i].iov_base;
-            sparse->areas[i].size = mmap_areas->areas[i].iov_len;
+            struct iovec *iov = &vfu_reg->mmap_areas[i];
+
             vfu_log(vfu_ctx, LOG_DEBUG, "%s: area %d %#llx-%#llx", __func__,
-                    i, sparse->areas[i].offset,
-                    sparse->areas[i].offset + sparse->areas[i].size);
+                    i, iov->iov_base, iov->iov_base + iov->iov_len);
```
Comment (swapnili) [inline lib/libvfio-user.c:157]: Looks like this should be iov->iov_base + iov->iov_len - 1 ?

**Review by swapnili — APPROVED** [review]
Looks good. Some minor nits.

File: include/libvfio-user.h [hunk 2]
```diff
@@ -241,6 +222,36 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
...
+ *    are handled by the library
+ *  - if no callback is provided, reads to other areas are a simple memcpy(),
+ *    and writes are an error
+ *  - otherwise, the callback is expected to handle the access
+ *
+ * @vfu_ctx: the libvfio-user context
+ * @region_idx: region index
+ * @size: size of the region
+ * @region_access: callback function to access region
+ * @flags: region flags (VFU_REGION_FLAG_)
```
[the remark this answers is not in the document]
Comment (jlevon) [reply include/libvfio-user.h:246]: done

File: lib/libvfio-user.c (hunk 1 above)
Comment (jlevon) [reply lib/libvfio-user.c:157]: sure, thanks

Comment (jlevon) [reply lib/libvfio-user.c:157]: actually, I'm not sure it does - if iov_len is zero, the range certainly isn't iov_base to iov_base - 1

I'll add brackets to indicate the open/closed range instead

**Review by swapnili — APPROVED** [review]

**Pull request merged**